docs/proposals/service-catalog.md

```diff
@@ -0,0 +1,123 @@
+- adding a new Service Catalog resource into which users can publish services that others can easily find and consume
+- adding support for dynamic provisioning of resources from the catalog
+- making it easier to link Services (and their configuration data) to Pods
+
+## Service configuration data
+
+The only data currently associated with a service is a list of zero or more endpoints. Many services also have configuration data associated with them. These include things such as logins, passwords, and additional connection parameters. To tie configuration data to a service, we propose adding information to the Service type to be able to express a relationship between the Service and its relevant configuration data. This could be accomplished by adding new field(s) or by using annotations. Potential configuration data target reference types include Secrets and ConfigData.
+
+Just creating the reference association by itself does not accomplish much; to be useful, that data needs to be available to the processes executing in containers. More on this process is below in the sections on service claims and linking.
+
+## Service catalogs
+
+A service catalog is a listing of "services". Examples might include an external database that is deployed outside of the cluster, an actual Kubernetes Service, a template of cluster primitives to be created, and custom provisioners that can perform tasks such as creating users in external systems.
+
+The catalog is not meant to include every service in the cluster. Instead, it should contain those services that users with to highlight and make available to other users. For example, your namespace might contain "etcd", "etcd-discovery", and "postgresql" services, and the only one you want to share with others is the postgresql service.
+
+### Publishing to a catalog
+
+Users should be able to publish entries to a service catalog. One means of accomplishing could be by adding annotations to the resources that can go into a catalog (e.g, Services). Another option would be to create `ServiceCatalogEntry` resources that reference the appropriate resources.
+
+An entry in the service catalog has a "type", which indicates the behavior that occurs when it is claimed by a user. We have thought of the following potential types:
+
+- reference: "I want to use this entry as-is, including its configuration data"
+- template: "I want to create items from the specified template"
+- provision: "I want the creation to be goverened by some other entity that implements the 'service broker' HTTP interface"
+
+We see the type as an arbitrary `string`; one or more controllers could run to process each claim type, performing whatever logic is appropriate to fulfill the specific type in question.
+
+### Viewing a catalog
+
+Users should be able to list the entries in a service catalog. Users should be able to select an entry and "consume" it. We call this consumption "claiming" a service from a service catalog.
```

Comment:
Commenting as I read, but one issue I struggle with is the claim pattern today implies exclusivity to the claimed resource which I do not think is the case in this discussion.  Need to think of a better metaphor.